polygeist/test/polygeist-opt/uniformity.mlir

```diff
@@ -53,3 +54,48 @@ func.func @test1(%arg0 : i1, %arg1: memref<?x!sycl_nd_item_2>)  {
   return
 }
 
+// COM: Check the uniformity for operations that load from memory.
+func.func @test2(%cond: i1, %val: i64, %arg1: memref<?x!sycl_nd_item_2>)  {
+  %alloca = memref.alloca() : memref<10xi64>
+
+  // COM: load memref arg. is uniform, index is uniform -> result is uniform.
+  // CHECK: test2_load1, uniformity: uniform  
+  %c0 = arith.constant 0 : index    
+  %load1 = memref.load %alloca[%c0] { tag = "test2_load1" } : memref<10xi64>
+
+  // COM: load memref arg. is uniform, index is non-uniform -> result is non-uniform.
+  // CHECK: test2_load2, uniformity: non-uniform
+  %c0_i32 = arith.constant 0 : i32  
+  %tx = sycl.nd_item.get_global_id(%arg1, %c0_i32) : (memref<?x!sycl_nd_item_2>, i32) -> i64
+  %tx_cast = arith.index_cast %tx : i64 to index
+  %load2 = memref.load %alloca[%tx_cast] { tag = "test2_load2" } : memref<10xi64>
+
+  // COM: load memref arg. uniformity is unknown due to store op. -> result uniformity is unknown.
+  // CHECK: test2_load3, uniformity: unknown
+  memref.store %val, %alloca[%c0]: memref<10xi64>
+  %load3 = memref.load %alloca[%c0] { tag = "test2_load3" } : memref<10xi64>
+
+  // COM: Stores in then/else blocks kill previous def. of %alloca and store uniform values.
+  %c1 = arith.constant 1 : i64
+  %c2 = arith.constant 2 : i64
+  scf.if %cond {
+    memref.store %c1, %alloca[%c0]: memref<10xi64>
+  } else {
+    memref.store %c2, %alloca[%c0]: memref<10xi64>
+  }
+  // COM: load memref arg. is uniform, index is uniform -> result is uniform.
+  // CHECK: test2_load4, uniformity: uniform
+  %load4 = memref.load %alloca[%c0] { tag = "test2_load4" }: memref<10xi64>
```

Comment:
should result be unknown here, as %cond is unknown.

Reply:
Right if the condition is not uniform then the thread would diverge and store either c1 or c2 end therefore the value stored to memory would be non-uniform. Good catch !

Reply:
Fixed.